### DISCUSSION TITLE
Cannot work with multiple model key types.

### DISCUSSION THREAD

**Discussion by irealworlds** [spatie/laravel-medialibrary#3745] [Ideas]
As it currently stands, it is not possible to use this library (out of the box) when working with models with multiple key types (i.e. numeric ids on some models, but uuids on others).

I am using UUIDs for most of my models, so I modified the Media table migration to use uuids for `model_id`, but some use integers. 

I could not find any way to achieve this other than a hacky solution like this, which is not DB agnostic and involves creating a custom trait
```php
trait InteractsWithMediaUsingNumericKey
{
    public function media(): MorphMany
    {
        return $this->morphMany(
            $this->getMediaModel(),
            'model',
            localKey: DB::raw('CAST(id AS VARCHAR)')->getValue(
                DB::getQueryGrammar(),
            ),
        );
    }
}
```

The best solution I could think of is changing `model_id` to a `json` column and using a json representation of the actual model id for all models, but this requires a change in the `HasMedia` contract as well as the `InteractsWithMedia` trait, because then the relation returned by `media()` is no longer `MorphMany`, but a `HasMany`.

**Comment by chrispage1** [comment]
I think this is a problem you'd get regardless. Generally, I'd say its always best practice to have an ID for a record, and then if you want to use UUID's have a secondary column that stores the UUID.

**Comment by ElonSmithee** [comment]
Лучший вариант, в данном случае, это использование разных моделей/таблиц, что я пытаюсь сейчас сделать, но по какой-то причине, пакет не особо волнует, что я тут добавил. 
```php
use Spatie\MediaLibrary\MediaCollections\Models\Media;

class OtherMedia extends Media
{
    protected $table = 'other_media';
}
```
```php
public function media(): MorphMany
{
    return $this->morphMany(OtherMedia::class, 'model');
}
```
```log
Data truncated for column 'model_id' at row 1 (Connection: mysql, SQL: insert into `media`...
```

**Comment by doncadavona** [comment]
Got into this issue, where my `\App\Models\User` uses ULID as primary key.

I fixed it by updating the to `$table->ulidMorphs('model');` from `$table->morphs('model');` in the migration file.

Of course for UUID, it would be:
`$table->uuidMorphs('model');` from `$table->morphs('model');`

Also, to keep it consistent, I use ULID or UUID on all of my tables, that may work for this issue, which has to handle multiple key types.

**Comment by patrasq** [comment]
Bump. Any proper fix?
